build-tools/src/main/kotlin/org/jetbrains/kotlin/bitcode/CompileToBitcode.kt

```diff
@@ -75,33 +80,41 @@ open class CompileToBitcode @Inject constructor(@InputDirectory val srcRoot: Fil
             return commonFlags + languageFlags + compilerArgs
         }
 
+    @get:SkipWhenEmpty
+    @get:InputFiles
     val inputFiles: Iterable<File>
         get() {
-            val srcFilesPatterns =
-                when (language) {
-                    Language.C -> listOf("**/*.c")
-                    Language.CPP -> listOf("**/*.cpp", "**/*.mm")
-                }
-            return project.fileTree(srcDir) {
-                it.include(srcFilesPatterns)
-                it.exclude(excludeFiles)
-            }.files
+            return srcDirs.flatMap { srcDir ->
+                project.fileTree(srcDir) {
+                    it.include(includeFiles)
+                    it.exclude(excludeFiles)
+                }.files
+            }
+        }
+
+    @get:InputFiles
+    protected val headers: Iterable<File>
+        get() {
+            return headersDirs.files.flatMap { dir ->
+                project.fileTree(dir) {
+                    it.include("**/*.h", "**/*.hpp")
```

Comment:
Not critical for now. But for C including `.hpp` files seems to be not fully right decision. Now we have no directories with mixed files, so it's ok.

Reply:
I've added a switch based on `language` of the task. Please take a look.